source/opt/decoration_manager.cpp

```diff
@@ -231,29 +231,62 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, true)) {
+    switch (inst->opcode()) {
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorate:
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
+    // TODO: update def-use manager (additional use of |to| )
+    switch (inst->opcode()) {
+      case SpvOpGroupDecorate:
+        // add |to| to list of decorated id's
+        inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
+        id_to_decoration_insts_[to].push_back(inst);
+        break;
+      case SpvOpGroupMemberDecorate:
+        // for each (id == from), add (to, literal) as operands
+        for (auto op = ++inst->begin(); op != inst->end(); ++op) { // emit decoration group id
+          if (op->words[0] == from) { // add new pair of operands: (to, literal)
+            inst->AddOperand(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
```

Comment:
This looks like a bug. std::vector::push_back may or may not invalidate |op| depending on if the vector has enough capacity.

Reply:
You mean if the allocation fails?

Reply:
The problem is that you are modifying a vector which you are iterating over. The C++ standard doesn't guarantee that you can do that.
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf
§ 23.3.7.5 vector modifiers
...
void push_back(const T& x);
void push_back(T&& x);
Remarks: Causes reallocation if the new size is greater than the old capacity. If no reallocation happens, all the iterators and references before the insertion point remain valid.
...

The standard guarantees that iterators remain valid only if vector has enough capacity. So there is a chance that after AddOperand |op| is not a valid iterator anymore.

Reply:
Thanks for clarifying!
So, I better create a new vector and append afterwards? (Already changed operand to &&.)

Reply:
I think you can still the same process if you use the vector indirectly with functions
GetNumOperands() and GetOperand(). But appending afterwards also works.